File: Java/백준/Silver/4335. 숫자 맞추기/숫자 맞추기.java
import java.io.IOException;

import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.io.BufferedWriter;

import java.lang.*;

public class Main {
    public static void main(String[] args) throws IOException {
        InputStreamReader reader = new InputStreamReader(System.in);
        BufferedReader br = new BufferedReader(reader);

        OutputStreamWriter writer = new OutputStreamWriter(System.out);
        BufferedWriter bw = new BufferedWriter(writer);

        ArrayList<Integer> inpList = new ArrayList<>();
        ArrayList<Boolean> highlowList = new ArrayList<>();

        int counter = 0;
        while (true) {
            int N = Integer.parseInt(br.readLine());
            if (N == 0) break;
            
            String inp = br.readLine();

            if (inp.equals("too high")) {
                counter += 1;
                inpList.add(N);
                highlowList.add(true);
            } else if (inp.equals("too low")) {
                counter += 1;
                inpList.add(N);
                highlowList.add(false);
            } else {
                String ans = "Stan may be honest";
                for (int i = 0; i < counter; i++) {
                    int num = inpList.get(i);
                    boolean flag = highlowList.get(i);
                    if (flag) {
                        if (num <= N) {
                            ans = "Stan is dishonest";
                            break;
                        }
                    } else {
                        if (num >= N) {
                            ans = "Stan is dishonest";
                            break;
                        }
                    }
                }
                System.out.println(ans);

                inpList.clear();
                highlowList.clear();
                counter = 0;
            }
        }
    }
}

File: Python/백준/Silver/4335. 숫자 맞추기/숫자 맞추기.py
import sys
input = sys.stdin.readline

limit = [0, 10]
ans = ("Stan is dishonest", "Stan may be honest")
honest = 1
while True:
    inp = input().rstrip()
    try:
        n = int(inp)
        if n == 0:
            break
    except:
        if inp == "right on":
            if not n in list(range(limit[0], limit[1] + 1)):
                honest = 0
            print(ans[honest])
            limit = [0, 10]
            honest = 1
            continue
        if honest == 0:
            continue
        else:
            if inp == 'too high':
                limit[1] = min(limit[1], n - 1)
            elif inp == 'too low':
                limit[0] = max(limit[0], n + 1)
            if limit[1] < limit[0]:
                honest = 0